<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [2]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [3]:
# load in the appropriate path to the file you want to load in
rockfile = "/Users/theodoreplotkin/desktop/postmalone/GA_Data_Science/DAT-06-24/class material/Unit 2/data/rock.csv"


In [4]:
# Load the data.
rock = pd.read_csv(rockfile)

In [5]:
# Look at the information regarding its columns.
rock.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [6]:
rock.columns

Index(['Song Clean', 'ARTIST CLEAN', 'Release Year', 'COMBINED', 'First?',
       'Year?', 'PlayCount', 'F*G'],
      dtype='object')

In [86]:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song         2230 non-null object
Artist       2230 non-null object
Release      2230 non-null object
Combined     2230 non-null object
First        2230 non-null int64
Year         2230 non-null int64
Playcount    2230 non-null int64
F*G          2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [8]:
# Change the column names when loading the '.csv':

col_names = ["Song", "Artist", "Release", "Combined", "First", "Year", "Playcount", "F*G"]

rock = pd.read_csv(rockfile, header = 0, names = col_names)
rock.head()

Song       Artist Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                Combined  First  Year  Playcount  F*G  
0        Caught Up in You by .38 Special      1     1         82   82  
1            Fantasy Girl by .38 Special      1     0          3    0  
2         Hold On Loosely by .38 Special      1     1         85   85  
3  Rockin' Into the Night by .38 Special      1     1         18   18  
4              Art For Arts Sake by 10cc      1     1          1    1

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [10]:
# Change the column names using the `.rename()` function.
rock = pd.read_csv(rockfile)
rock.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [17]:
#recall the .rename method takes a dictionary as input 
rock.rename(columns={"Song Clean":"Song",
                     "ARTIST CLEAN":"Artist",
                     "Release Year":"Release", 
                     "COMBINED":"Combined",
                     "First?":"First",
                     "Year?":"Year",
                     "PlayCount":"Playcount",
                     "F*G":"F*G"}, inplace = True)
rock.head()

Song       Artist Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                Combined  First  Year  Play Count  F*G  
0        Caught Up in You by .38 Special      1     1          82   82  
1            Fantasy Girl by .38 Special      1     0           3    0  
2         Hold On Loosely by .38 Special      1     1          85   85  
3  Rockin' Into the Night by .38 Special      1     1          18   18  
4              Art For Arts Sake by 10cc      1     1           1    1

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [7]:
# Replace the column names by reassigning the `.columns` attribute.

In [18]:
rock = pd.read_csv(rockfile)
rock.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [30]:
col_names = ["Song", "Artist", "Release", "Combined", "First", "Year", "Playcount", "F*G"]

rock.columns = col_names
rock.head()

Song       Artist Release  \
0        Caught Up in You  .38 Special    1982   
1            Fantasy Girl  .38 Special     NaN   
2         Hold On Loosely  .38 Special    1981   
3  Rockin' Into the Night  .38 Special    1980   
4       Art For Arts Sake         10cc    1975   

                                Combined  First  Year  Playcount  F*G  
0        Caught Up in You by .38 Special      1     1         82   82  
1            Fantasy Girl by .38 Special      1     0          3    0  
2         Hold On Loosely by .38 Special      1     1         85   85  
3  Rockin' Into the Night by .38 Special      1     1         18   18  
4              Art For Arts Sake by 10cc      1     1          1    1

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [57]:
# Show records where df['release'] is null
rock["Release"].isnull().head()

0    False
1     True
2    False
3    False
4    False
Name: Release, dtype: bool

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [89]:
rock.Release.fillna(value = 0, inplace = True)
#note we include the inplace = True to overwrite the changes to the dataframe
rock.head(100)

Song                Artist Release  \
0                      Caught Up in You           .38 Special    1982   
1                          Fantasy Girl           .38 Special       0   
2                       Hold On Loosely           .38 Special    1981   
3                Rockin' Into the Night           .38 Special    1980   
4                     Art For Arts Sake                  10cc    1975   
5                            Kryptonite          3 Doors Down    2000   
6                                 Loser          3 Doors Down    2000   
7                         When I'm Gone          3 Doors Down    2002   
8                            What's Up?         4 Non Blondes    1992   
9                            Take On Me                  a-ha    1985   
10                Baby, Please Don't Go                 AC/DC       0   
11                        Back In Black                 AC/DC    1980   
12                              Big Gun                 AC/DC    1993   
13               CAN'T STOP ROCK'N'ROLL                 AC/DC       0   
14          Dirty Deeds Done Dirt Cheap                 AC/DC    1976   
15              For Those About To Rock                 AC/DC    1981   
16                     Girls Got Rhythm                 AC/DC       0   
17                       Hard As A Rock                 AC/DC    1995   
18                   Have a Drink On Me                 AC/DC    1980   
19                          Hells Bells                 AC/DC    1980   
20                      Highway To Hell                 AC/DC    1979   
21           It's A Long Way To The Top                 AC/DC    1975   
22                            Jailbreak                 AC/DC    1984   
23                    Let There Be Rock                 AC/DC    1977   
24                      Let's Get It Up                 AC/DC       0   
25                            Live Wire                 AC/DC       0   
26                           Moneytalks                 AC/DC       0   
27                        Night Prowler                 AC/DC    1979   
28  Rock and Roll Ain't Noise Pollution                 AC/DC    1980   
29                      Shoot To Thrill                 AC/DC       0   
..                                  ...                   ...     ...   
70                        What It Takes             Aerosmith    1989   
71                    You See Me Crying             Aerosmith       0   
72                    Hand In My Pocket     Alanis Morissette    1995   
73                      You Oughta Know     Alanis Morissette    1995   
74                         Black Velvet         Alannah Myles    1989   
75                              Fantasy             Aldo Nova       0   
76                          BE MY LOVER          Alice Cooper       0   
77                Billion Dollar Babies          Alice Cooper    1973   
78                             Eighteen          Alice Cooper    1971   
79                              Elected          Alice Cooper    1973   
80                        House Of Fire          Alice Cooper       0   
81                         I'm Eighteen          Alice Cooper    1971   
82                 No More Mr. Nice Guy          Alice Cooper    1973   
83                         School's Out          Alice Cooper    1972   
84                      Under My Wheels          Alice Cooper    1971   
85              Welcome To My Nightmare          Alice Cooper    1975   
86                       Down in a Hole       Alice In Chains    1992   
87                       Man In The Box       Alice In Chains    1990   
88                           No Excuses       Alice In Chains    1994   
89                              Rooster       Alice In Chains    1992   
90                           Them Bones       Alice In Chains    1992   
91                               Would?       Alice In Chains    1992   
92           Ain't Wastin' Time No More  Allman Brothers Band    1972   
93                  

In [90]:
#a useful tip for replacing NaNs in more than one column at a time

#the .select_dtypes() method, takes include = <np.some data type> as an argument 
cat_columns = rock.select_dtypes(include = np.object).columns.tolist()

In [92]:
rock[cat_columns] #would select only the columns which include categorical data 

Song         Artist Release  \
0                                Caught Up in You    .38 Special    1982   
1                                    Fantasy Girl    .38 Special       0   
2                                 Hold On Loosely    .38 Special    1981   
3                          Rockin' Into the Night    .38 Special    1980   
4                               Art For Arts Sake           10cc    1975   
5                                      Kryptonite   3 Doors Down    2000   
6                                           Loser   3 Doors Down    2000   
7                                   When I'm Gone   3 Doors Down    2002   
8                                      What's Up?  4 Non Blondes    1992   
9                                      Take On Me           a-ha    1985   
10                          Baby, Please Don't Go          AC/DC       0   
11                                  Back In Black          AC/DC    1980   
12                                        Big Gun          AC/DC    1993   
13                         CAN'T STOP ROCK'N'ROLL          AC/DC       0   
14                    Dirty Deeds Done Dirt Cheap          AC/DC    1976   
15                        For Those About To Rock          AC/DC    1981   
16                               Girls Got Rhythm          AC/DC       0   
17                                 Hard As A Rock          AC/DC    1995   
18                             Have a Drink On Me          AC/DC    1980   
19                                    Hells Bells          AC/DC    1980   
20                                Highway To Hell          AC/DC    1979   
21                     It's A Long Way To The Top          AC/DC    1975   
22                                      Jailbreak          AC/DC    1984   
23                              Let There Be Rock          AC/DC    1977   
24                                Let's Get It Up          AC/DC       0   
25                                      Live Wire          AC/DC       0   
26                                     Moneytalks          AC/DC       0   
27                                  Night Prowler          AC/DC    1979   
28            Rock and Roll Ain't Noise Pollution          AC/DC    1980   
29                                Shoot To Thrill          AC/DC       0   
...                                           ...            ...     ...   
2200                      Owner of a Lonely Heart            Yes    1983   
2201                                   Roundabout            Yes    1971   
2202                             Starship Trooper            Yes    1971   
2203                            Wonderous Stories            Yes       0   
2204                         Yours Is No Disgrace            Yes    1971   
2205                        Tell Me What You Want          Zebra       0   
2206                        Who's Behind The Door          Zebra       0   
2207                           Time of the Season        Zombies    1968   
2208                    A Fool for Your Stockings         ZZ Top       0   
2209             Arrested For Driving While Blind         ZZ Top    1976   
2210                 Beer Drinkers & Hell Raisers         ZZ Top    1973   
2211                             Cheap Sunglasses         ZZ Top    1980   
2212                         Gimme All Your Lovin         ZZ Top    1983   
2213                        Got Me Under Pressure         ZZ Top    1983   
2214                            Heard It On the X         ZZ Top    1975   
2215                                  I Thank You         ZZ Top       0   
2216                      I'm Bad, I'm Nationwide         ZZ Top       0   
2217                      Jesus Just Left Chicago         ZZ Top    1973   
2218                                Just Got Paid         ZZ Top       0   
2219                                    La Grange         ZZ Top    1973   
2220                                         Legs         ZZ Top    1983   
2221             

#### 4.B Verify that `release` contains no null values.

In [70]:
# A:
rock.Release.isnull().sum()

0

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [80]:
# A:
rock.dtypes

#Release should be an int64 and not a string!

Song         object
Artist       object
Release      object
Combined     object
First         int64
Year          int64
Playcount     int64
F*G           int64
dtype: object

### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [87]:
# A:
i = 0
empty = []
for year in rock["Release"]:
    if type(year) == str:
        empty.append(i)
    i += 1
len(empty)

1653

In [93]:
rock["Release"].unique()

array(['1982', 0, '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM', '1963',
       '1961'], dtype=object)

In [97]:
mask = (rock.Release == "Release Year") | (rock.Release == "SONGFACTS.COM")

rock.loc[mask, "Release"] = 0
#writes over the row where release is "SONGFACTS.COM" with the value 0 

Series([], Name: Release, dtype: object)

In [107]:
#rock.isnull().sum() > 0

#rock.loc[:, rock.isnull().sum() > 0 ] 

Song         False
Artist       False
Release      False
Combined     False
First        False
Year         False
Playcount    False
F*G          False
dtype: bool

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [13]:
# A:

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [14]:
# A:

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [112]:
# A:
rock["Release"].describe()

count     2230
unique      56
top          0
freq       578
Name: Release, dtype: int64

In [113]:
max(rock["Release"])

TypeError: '>' not supported between instances of 'int' and 'str'

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

In [114]:
# A:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song         2230 non-null object
Artist       2230 non-null object
Release      2230 non-null object
Combined     2230 non-null object
First        2230 non-null int64
Year         2230 non-null int64
Playcount    2230 non-null int64
F*G          2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


In [116]:
#the .astype() method allows us to loop thru an entire column and change its type
    #i.e. np.int16 for 16-bit integer

rock["Release"] = rock["Release"].astype(np.int16)
#also note that for large datasets, the bit-size of your entires really matter alot 

#8-bit is too small for release year since 2 ** 8 = 256 
#16-bit works well for release date since release years contain 4 values 
#64-bit is overkill for this situation 

In [117]:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song         2230 non-null object
Artist       2230 non-null object
Release      2230 non-null int16
Combined     2230 non-null object
First        2230 non-null int64
Year         2230 non-null int64
Playcount    2230 non-null int64
F*G          2230 non-null int64
dtypes: int16(1), int64(4), object(3)
memory usage: 126.4+ KB


In [118]:
rock["Release"].describe()

count    2230.000000
mean     1465.331390
std       867.196161
min         0.000000
25%         0.000000
50%      1973.000000
75%      1981.000000
max      2014.000000
Name: Release, dtype: float64

_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [130]:
# A:
def print_info(row):
    return row["Song"], row["Artist"] , row["Release"] < 1970

print_info(rock.loc[4,:])

('Art For Arts Sake', '10cc', False)

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.